### ISSUE TITLE
Ordering by a boolean column twice after an outer join results in invalid SQL

### ISSUE THREAD

**Issue by neoGeneva** [aspnet/EntityFrameworkCore#12249]
A query that has two order bys on the same column, after an outer join, on a column that is a boolean, results in invalid SQL.

The other scenarios I've tested are inner join, and ordering by a int column, neither generate the same error.

This issue exists in 2.1.0 and 2.2.0-preview1-34386

```
Exception message: System.Data.SqlClient.SqlException (0x80131904): A column has been specified more than once in the order by list. Columns in the order by list must be unique.
Stack trace:
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlDataReader.TryConsumeMetaData()
   at System.Data.SqlClient.SqlDataReader.get_MetaData()
   at System.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString)
   at System.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, Boolean async, Int32 timeout, Task& task, Boolean asyncWrite, SqlDataReader ds)
   at System.Data.SqlClient.SqlCommand.ExecuteReader(CommandBehavior behavior)
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.Execute(IRelationalConnection connection, DbCommandMethod executeMethod, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.ExecuteReader(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.BufferlessMoveNext(DbContext _, Boolean buffer)
   at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.MoveNext()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider._TrackEntities[TOut,TIn](IEnumerable`1 results, QueryContext queryContext, IList`1 entityTrackingInfos, IList`1 entityAccessors)+MoveNext()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext()
```

### Steps to reproduce
Here's an example project that demonstrates the issue: [EFCoreBugs.OrderByDuplicate.zip](https://github.com/aspnet/EntityFrameworkCore/files/2074223/EFCoreBugs.OrderByDuplicate.zip)

Following is code that throws the excption:

```c#
var queryThrowsSqlException = db.Cart
    .Select(c => c.DeliveryAddress)
    .OrderByDescending(d => d.IsDefault)
    .OrderByDescending(d => d.IsDefault)
    .ToArray();
```

```c#
public class Cart
{
    public int CartId { get; set; }
    public int? DeliveryAddressId { get; set; }
    public DeliveryAddress DeliveryAddress { get; set; }
    public int MemberId { get; set; }
    public Member Member { get; set; }
}

public class DeliveryAddress
{
    public int DeliveryAddressId { get; set; }
    public bool IsDefault { get; set; }

    public ICollection<Cart> Carts { get; set; }
}
```

### Further technical details
EF Core version: 2.1.0 (still exists in 2.2.0-preview1-34386)
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10 Version 1709 (OS Build 16299.431)
IDE: Visual Studio 2017 15.7.3

**Comment by maumar** [comment]
We produce the following query model:

``cs`
from Cart c in DbSet<Cart>
join DeliveryAddress c.DeliveryAddress in DbSet<DeliveryAddress>
on Property([c], "DeliveryAddressId") equals (Nullable<int>)Property([c.DeliveryAddress], "DeliveryAddressId") into c.DeliveryAddress_group
from DeliveryAddress c.DeliveryAddress in 
    (from DeliveryAddress c.DeliveryAddress_groupItem in [c.DeliveryAddress_group]
    select [c.DeliveryAddress_groupItem]).DefaultIfEmpty()
order by (bool)?[c.DeliveryAddress] | ?[c.DeliveryAddress] | [c.DeliveryAddress]?.IsDefault == True? == True? desc
order by (bool)?[c.DeliveryAddress] | ?[c.DeliveryAddress] | [c.DeliveryAddress]?.IsDefault == True? == True? desc
select [c.DeliveryAddress]
``

Ordering is so complex because we introduce safeguards due to type differences - since c.DeliveryAddress can be null, it's IsDefault property could also end up null, but the result is expected to be non-nullable bool. 

Problem is that we are not able to recognize expressions in that form to be the same.

**Comment by maumar** [comment]
fixed in 9e665c08cdb5c44084ddaae63cdd9c0ff043784d
